Playwright: Abstract Handler Creation for MSW

ardelato opened the pull request:
## Description

In #1275 we created the ability to set dynamic handlers for `MSW`. However, the API for creating a handler is rather bloated and a lot of it can be extracted to reduce the boilerplate.

As such, we created a new static helper function to create these handlers.

## CR:

Commit by commit would be the easiest way to understand the code as some obscure context is needed in some cases.

Overall, the end goal was to abstract creating these dynamic handlers just enough to reduce the boilerplate and still be simple to use. Most of this was accomplished by making use of Typescript's features. 

## QA:

Besides checking CI still passing, I am not sure if it will be worth validating if the error handling works, as that might be exhaustive. 

I already confirmed that the `international buybox` handler still works using similar steps as https://github.com/iFixit/react-commerce/pull/1319#issuecomment-1416451999.

I also confirmed that the type checking is working as expected for the overloaded signatures:


|Before|After|
|:---:|:---:|
|  ![GraphQL body type is allowed](https://user-images.githubusercontent.com/22064420/217672866-e36d2476-6155-44df-ac17-bea9889c653b.png) | ![Graphql only accepts an object not a string](https://user-images.githubusercontent.com/22064420/217672913-be2a3a05-58be-4f01-87e1-c6abc59485db.png)|
<p align=center> <strong>In the scenario above, a GraphQLHandler requires a body that is only of type <code>Object</code></strong></p>


cr_req 2
qa_req 0

Connects: #1290

mlahargou commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -82,37 +115,45 @@ export default class Handler {
-            (req, res, ctx) => {
-               if (passthrough) return req.passthrough();
-               if (once) return res.once(ctx.status(status), ctx.data(body));
+            customResolver ??
```
When do you imagine we would use a custom resolver?

ardelato replied:
I could see it being used when we want to use the values of the `request` to modify the `response` or reduce the number of handler creations needed.

For example, if wanted to validate each user type in a single test, we would need to create each handler that would override the same path.

Instead, we could bake each case into a single custom resolver:

```typescript
server.use(Handler.create({endpoint: '/user', method: 'get'}, pro-test-user-response))
// Run test

server.resetHandlers()
Handler.create({endpoint: '/user', method: 'get'}, ifixit-empl-user-response)
// Run test

server.resetHandlers()
...

// vs.

const customR = (req, res, ctx) => {
	const user = req.body.username

	if(username = 'pro-test-user') {
		return res(
			ctx.json({...defaultUserProps, discountTier: ''})	
	}
	else if (username = 'ifixt-empl-test-user') {}
	else if(username = 'reg-test-user') {}
	else 
	...
}

Handler.create({endpoint: '/user', method: 'get'}, customR)
```

ardelato replied:
Obviously, we could just separate each `user` test into its own `test case` but I am just trying to give an example of how the `customResovler` option could be used.

mlahargou replied:
oh cool! That makes sense 👍

mlahargou reviewed (approved):
CR 📱 

Nice job on this! I think it makes mocking these requests much simpler.

andyg0808 commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -0,0 +1,233 @@
+ * @see https://mswjs.io/docs/api/response
+ *
+ */
+export type MockedResponseInfo = {
+   status: number;
+   headers?: Record<string, string | string[]>;
+   body?: Record<string, any> | string | number | boolean | null;
+   responseType?: 'json' | 'text' | 'xml' | 'raw';
+};
```
I'd suggest splitting this into two types: one for GraphQL requests and one for REST requests. I think trying to keep them together and then filter the appropriate keys out below is probably not simplifying things. If you want to avoid duplication on `status`, you can always create a `type BaseResponseInfo = { status: number }` and then intersect it into the field set for each subtype.

andyg0808 commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -0,0 +1,233 @@
+   DefaultBodyType
+>;
+
+export type RequestInfo = {
+   endpoint: string;
+   method: GraphQLMethods | RestMethods;
+};
```
I think it'll be easier to work with to have two copies of this: one for GraphQL and one for REST. The fact that `endpoint` is the same type and would be duplicated seems relatively unimportant to me. If you disagree, you can always define a `BaseRequestInfo` with just `{ endpoint: string }` in it.

ardelato replied:
If we make two respective request info types for each API type, should we remove the overloading for `create()`?

andyg0808 commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -0,0 +1,233 @@
+      const { once, passthrough, customResolver } = options;
+
+      if (isGraphQLMethod(method)) {
+         if (!isRecord(body)) {
+            throw new Error(`Invalid GraphQL body ${body}. Must be an object`);
+         }
```
I'm guessing that this check can be made redundant with enough care in typing. I'm not opposed to leaving it even in that case, in order to provide safety against untyped calls, but if TypeScript forced this check, it's probably indicative of a deeper problem.

ardelato replied:
It became redundant in [`cd6fcf4` (#1348)](https://github.com/iFixit/react-commerce/pull/1348/commits/cd6fcf42ebeae5708a4edfa0958a3c48987fe9dd). Prior to that, the `body` type for a GraphQL Handler could have been anything acceptable under a Rest `body` type. That is why the check was there but is no longer needed.

andyg0808 commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -0,0 +1,233 @@
+            '*',
+            customResolver ??
+               ((req, res, ctx) => {
+                  if (passthrough) return req.passthrough();
+                  if (once) return res.once(ctx.status(status), ctx.data(body));
```
Since `passthrough` and `once` don't do anything when `customResolver` is present, I might try to make the `HandlerOptions` type reflect that:
`type GraphQLHandlerOptions = { customResolver: GraphQLResolver } | { passthrough?: boolean; once: boolean; }`

andyg0808 reviewed (approved):
Lots of typey suggestions, but looks good overall.
deploy_block :+1: on comments.

ardelato commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -0,0 +1,233 @@
+ *
+ * @see https://mswjs.io/docs/api/rest#methods
+ */
+export type RestMethods = 'all' | 'get' | 'post' | 'put' | 'patch' | 'delete';
```
Unfortunately, there is no `MSWRestMethods` type but I can still accomplish this by doing

```typescript
Omit<keyof (typeof rest), 'options`>
```

That is definitely better then the static typing so I'll make the change 👍

ardelato commented:
un_dev_block ⚡ un_deploy_block ⚡ I addressed the comments by exporting two different functions instead of a single class for creating each handler type.

andyg0808 commented on frontend/tests/playwright/msw/request-handler.ts:
```diff
@@ -86,148 +110,121 @@ export type MockedResponseInfo = {
+   const { endpoint, method } = request;
+   const { status, body } = response;
+
+   const customResolver =
+      typeof options === 'function' ? (options as GraphQLResolver) : null;
+   const { once = false, passthrough = false } =
+      typeof options === 'object' ? (options as HandlerOptions) : {};
```
Not a big deal, but have you tried leaving these casts out? I'd have expected TypeScript to potentially be able to infer the types here, and doing that would probably be safer for the long-term.

ardelato replied:
Unfortunately, Typescript does not implicitly understand that `once` and `passthrough` exists in `options`

![image](https://user-images.githubusercontent.com/22064420/219523640-1b799eb6-152d-4cf9-87f7-d6fa11aeeeda.png)

And `customResovler` would resolve to `null` instead of either `RestResolver` or `null`

![image](https://user-images.githubusercontent.com/22064420/219523761-1b310aa9-2a80-4255-8f8b-b6d15d351d92.png)

andyg0808 reviewed (approved):
CR :+1:

mlahargou reviewed (approved):
CR 📱 Sweet!
